### PR TITLE
Add exceptions to _load_config_from_filesystem for no store key or no config params

### PR DESCRIPTION
**Pull request by rgadon107**
@hellofromtonya I'm trying to test edge case #5: 

>When there’s no key in the configuration array that was loaded from the filesystem, the values for:
```list( $store_key, $config ) = _load_config_from_filesystem( $path_to_file );```

>`$store_key` will be `NULL`

>and if the array is empty, then `$config` will be `False`.

>If there’s no key, something else should happen to alert that there was no key in the configuration file.

Choices:

>1. Throw an error.
>2. Return something like `False` or `Null`."

Trying to throw an \Exception from `_load_config_from_filesystem`. 

Message: Failed asserting that exception of type "Exception" is thrown.

I'm trying to avoid using a test fixture inside `_load_config_from_filesystem` and mock the function parameter `$path_to_file` with Brain Monkey instead.

### CODE REVIEW

File: mu-plugins/central-hub/src/config-store/internals.php [hunk 1]
```diff
@@ -62,13 +62,22 @@ function _the_store( $store_key = '', $config_to_store = array(), $remove = null
...
- *
  * @param string $path_to_file Absolute path to the config file.
  *
  * @return array
+ * @throws \Exception
+ * @since 1.1.0
+ *
  */
 function _load_config_from_filesystem( $path_to_file ) {
+	if ( ! $path_to_file ) {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/src/config-store/internals.php]: The Exception are thrown relates not to the path/to/file, but to what is returned from loading the file.  You want to check what is returned in `$config`.  In other words, you're checking what's inside of that configuration file.  More below....

Also, you don't need to check if `$path_to_file` is valid.  Why?  You are using a `require`.  [Look at the PHP manual](https://www.php.net/manual/en/function.require.php).  It says:

>`require` is identical to `include` except upon failure it will also produce a fatal `E_COMPILE_ERROR` level error.

If `$path_to_file` is empty or invalid, PHP will throw a fatal error when it tries to load.

File: mu-plugins/central-hub/src/config-store/internals.php [hunk 2]
```diff
@@ -62,13 +62,22 @@ function _the_store( $store_key = '', $config_to_store = array(), $remove = null
...
+		throw new \Exception(
+			sprintf(
+				'The configuration is empty and the store_key does not exist.'
+			)
+		);
+	}
+
 	$config = (array) require $path_to_file;
 
 	return array(
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/src/config-store/internals.php]: Let's revisit what you are trying to check for:

1. Empty key means no store key.
2. Empty config means nothing to store in the store.

How can we check that?  Here's one implementation:

```php
/**
 * Load a configuration from the filesystem, returning its storage key and configuration parameters.
 *
 * @since 1.0.0
 *
 * @param string $path_to_file Absolute path to the config file.
 *
 * @return array returns an array with storage key => config parameters.
 * @throws \Exception
 */
function _load_config_from_filesystem( $path_to_file ) {
	$config    = (array) require $path_to_file;

	$store_key = key( $config );
	if ( empty( $store_key ) ) {
		throw new \Exception(
			sprintf( 'No store key exists in the %s configuration file.', esc_attr( $path_to_file ) )
		);
	}

	$config_params = current( $config );
	if ( empty( $config_params ) ) {
		throw new \Exception(
			sprintf( 'No configuration parameters exist for store key [%s] in the %s configuration file.',
				esc_attr( $store_key ), esc_attr( $path_to_file ) )
		);
	}

	return [ $store_key, $config_params ];
}
```

File: mu-plugins/central-hub/src/config-store/internals.php (hunk 1 above)
Comment (rgadon107) [reply mu-plugins/central-hub/src/config-store/internals.php]: @hellofromtonya We _did_ talk several weeks back about the `require` control structure and how it would throw a fatal error if a value was not returned to the argument. That's why I thought the use of a conditional check on that argument would be appropriate. 

But you are right, it's not the argument `$path_to_file` that we're interested in. It's the contents of the configuration that needs checking.

File: mu-plugins/central-hub/src/config-store/internals.php (hunk 2 above)
Comment (rgadon107) [reply mu-plugins/central-hub/src/config-store/internals.php]: @hellofromtonya Elegant and practical! Also, you clarified something for me by this example. 

I was unsure whether to add a string placeholder ( [%s] ) to the Exception message. You just answered my question with your example.

File: mu-plugins/central-hub/src/config-store/internals.php [hunk 3]
```diff
@@ -62,13 +62,22 @@ function _the_store( $store_key = '', $config_to_store = array(), $remove = null
...
- *
  * @param string $path_to_file Absolute path to the config file.
  *
  * @return array
+ * @throws \Exception
+ * @since 1.1.0
+ *
  */
 function _load_config_from_filesystem( $path_to_file ) {
+	if ( is_null( $path_to_file ) ) {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/src/config-store/internals.php]: @rgadon107 This conditional and error is not needed.  It's redundant as PHP will throw an error automatically through the `require`.

File: mu-plugins/central-hub/src/config-store/internals.php (hunk 2 above)
Comment (rgadon107) [reply mu-plugins/central-hub/src/config-store/internals.php]: @hellofromtonya Code above incorporated into branch.

File: mu-plugins/central-hub/tests/phpunit/fixtures/config-with-params-only.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * @since      1.0.0
+ * @author     Robert A. Gadon
+ * @link       https://github.com/rgadon107/cornerstone
+ * @license    GNU General Public License 2.0+
+ */
+
+namespace spiralWebDb\centralHub\Tests\Fixtures;
+
+return  [
+		'aaa' => 'bbb',
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/fixtures/config-with-params-only.php]: @rgadon107 The configuration parameters should be an array in an array.  Right?  The structure of this file should be:

```php
[
   'store_key' => [
         // config parameters
   ],
]
```

Using that structure, this file should be:

```php

```php
return [
	[
		'aaa' => 'bbb',
		'ccc' => 'ddd',
	],
];
```

Does that make sense?

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php [hunk 5]
```diff
@@ -45,4 +46,36 @@ public function test_should_return_array_when_path_to_file_is_given() {
...
 	}
+
+	/**
+	 * Test _load_config_from_filesystem() should throw an Exception when configuration store key is empty
+	 *  and parameters exist.
+	 */
+	public function test_should_throw_exception_when_no_store_key() {
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage(
+			sprintf( 'No store key exists in the [path_to_file] configuration file.', 'path_to_file' )
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php]: @rgadon107 What's wrong here?  

You need to insert the path to file inside of the message's string.  You use `sprintf` to replace a placeholder with the value from a variable:

```php
sprint( 'some message %s', $some_variable )
```

So the placeholder is `%s` and it is replaced by the value associated with the variable.

Now, let's relate that concept to this edge case.

The message the [error will throw is found here](https://github.com/rgadon107/cornerstone/pull/20/files#diff-aff0ac70304193f0d42b37b100807912R77):

```php
function _load_config_from_filesystem( $path_to_file ) {
	$config = (array) require $path_to_file;
	$config    = (array) require $path_to_file;
	$store_key = key( $config );
	if ( empty( $store_key ) ) {
		throw new \Exception(
			sprintf( 'No store key exists in the %s configuration file.', esc_attr( $path_to_file ) )
		);
	}
```

Notice the structure of the message:

1. It needs a variable to replace out the placeholder of `%s`.
2. There's no square brackets around the path to file.

Put it all together and the message that will be thrown is:

```
No store key exists in the path/to/file configuration file.
```

Now you need to replace out that `path/to/file` with the right path to file in the test.  How do you do that?

1. Notice that `$path_to_file` is below this exception.  You need that variable.  Reorder by moving that line of code up to the top.
2. Fix the `sprintf`.

I fixed this in commit https://github.com/rgadon107/cornerstone/pull/20/commits/e1e68716f9375ac62efe610c8599180f4686b255

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php [hunk 6]
```diff
@@ -45,4 +46,39 @@ public function test_should_return_array_when_path_to_file_is_given() {
...
+
+	/**
+	 * Test _load_config_from_filesystem() should throw an Exception when configuration store key exists and
+	 *  parameters are empty.
+	 */
+	public function test_should_throw_exception_when_config_store_key_exists_and_no_parameters() {
+		$path_to_file = CENTRAL_HUB_ROOT_DIR . '/tests/phpunit/fixtures/config-with-store-key-only.php';
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage(
+			sprintf( 'No configuration parameters exist for store key [%s] in the %s configuration file.',
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php]: @rgadon107 For this `sprintf`, you don't need to pass the store key of `'foo'` as an argument.  Rather, put it into the message directly like this:

```php
sprintf( 'No configuration parameters exist for store key [foo] in the %s configuration file.', $path_to_file )
```

Why?  You know what the store key should be.  It's not a programmatic variable like the value in `$path_to_file`.  Right?  Therefore, you can insert it directly into the message.

Remember, `sprintf` is programmatically building a string by replacing out placeholders with the value from code, where code is a variable, function, math, etc.

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php [hunk 7]
```diff
@@ -45,4 +46,39 @@ public function test_should_return_array_when_path_to_file_is_given() {
...
+			sprintf( 'No store key exists in the [%s] configuration file.', $path_to_file )
+		);
+		_load_config_from_filesystem( $path_to_file );
+	}
+
+	/**
+	 * Test _load_config_from_filesystem() should throw an Exception when configuration store key exists and
+	 *  parameters are empty.
+	 */
+	public function test_should_throw_exception_when_config_store_key_exists_and_no_parameters() {
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php]: Let's shorten the method name like we did with the previous test:

```php

/**
 * Test _load_config_from_filesystem() should throw an Exception when the configuration has no 
 * configuration parameters.
 */
public function test_should_throw_exception_when_config_has_no_parameters()
```

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php (hunk 6 above)
Comment (hellofromtonya) [reply mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php]: Fixed with commit https://github.com/rgadon107/cornerstone/pull/20/commits/618020b287430994ad863bf1f240ade00863c83a

File: mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php (hunk 7 above)
Comment (hellofromtonya) [reply mu-plugins/central-hub/tests/phpunit/unit/config-store/_loadConfigFromFilesystem.php]: Fixed with https://github.com/rgadon107/cornerstone/pull/20/commits/618020b287430994ad863bf1f240ade00863c83a
